src/coreclr/jit/llvm.cpp

```diff
@@ -841,26 +852,131 @@ llvm::Value* Llvm::buildUserFuncCall(GenTreeCall* call)
+}
+
+bool Llvm::helperRequiresShadowStack(CORINFO_METHOD_HANDLE corinfoMethodHnd)
+{
+    return corinfoMethodHnd == _compiler->eeFindHelper(CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE);
+}
+
+void Llvm::buildHelperFuncCall(GenTreeCall* call)
+{
+    if (call->gtCallMethHnd == _compiler->eeFindHelper(CORINFO_HELP_READYTORUN_GENERIC_HANDLE) ||
+        call->gtCallMethHnd == _compiler->eeFindHelper(CORINFO_HELP_READYTORUN_GENERIC_STATIC_BASE) ||
+        call->gtCallMethHnd == _compiler->eeFindHelper(CORINFO_HELP_READYTORUN_DELEGATE_CTOR))
+    {
+        // TODO-LLVM
+        failFunctionCompilation();
+    }
+
+    if (call->gtCallMethHnd == _compiler->eeFindHelper(CORINFO_HELP_READYTORUN_STATIC_BASE))
+    {
+        const char* symbolName = (*_getMangledSymbolName)(_thisPtr, call->gtEntryPoint.handle);
+        Function* llvmFunc = _module->getFunction(symbolName);
+        if (llvmFunc == nullptr)
         {
-            const char* symbolName = (*_getMangledSymbolName)(_thisPtr, call->gtEntryPoint.handle);
-            Function* llvmFunc = _module->getFunction(symbolName);
-            if (llvmFunc == nullptr)
-            {
-                llvmFunc = Function::Create(FunctionType::get(Type::getInt8PtrTy(_llvmContext), ArrayRef<Type*>(Type::getInt8PtrTy(_llvmContext)), false), Function::ExternalLinkage, 0U, symbolName, _module); // TODO: ExternalLinkage forced as defined in ILC module
-            }
+            llvmFunc = Function::Create(buildHelperLlvmFunctionType(call, true), Function::ExternalLinkage, 0U, symbolName, _module); // TODO: ExternalLinkage forced as defined in ILC module
+        }
 
-            // replacement for _info.compCompHnd->recordRelocation(nullptr, gtCall->gtEntryPoint.handle, IMAGE_REL_BASED_REL32);
-            (*_addCodeReloc)(_thisPtr, call->gtEntryPoint.handle);
+        // replacement for _info.compCompHnd->recordRelocation(nullptr, gtCall->gtEntryPoint.handle, IMAGE_REL_BASED_REL32);
+        (*_addCodeReloc)(_thisPtr, call->gtEntryPoint.handle);
 
-            mapGenTreeToValue(node, _builder.CreateCall(llvmFunc, getShadowStackForCallee()));
-            return;
+        mapGenTreeToValue(call, _builder.CreateCall(llvmFunc, getShadowStackForCallee()));
+        return;
+    }
+    else
+    {
+        fgArgInfo* argInfo = call->fgArgInfo;
+        unsigned int argCount = argInfo->ArgCount();
+        fgArgTabEntry** argTable = argInfo->ArgTable();
+        std::vector<OperandArgNum> sortedArgs = std::vector<OperandArgNum>(argCount);
+        OperandArgNum* sortedData = sortedArgs.data();
+        bool requiresShadowStack = helperRequiresShadowStack(call->gtCallMethHnd);
+
+        for (unsigned i = 0; i < argCount; i++)
+        {
+            fgArgTabEntry* curArgTabEntry = argTable[i];
+            unsigned int   argNum = curArgTabEntry->argNum;
+            OperandArgNum  opAndArg = { argNum, curArgTabEntry->GetNode() };
+            sortedData[argNum] = opAndArg;
+        }
```

Comment:
It would be nice to share the logic between user and helper calls; it seems they have no fundamental reason to be handled differently. I understand currently the sharing would not achieve a lot (there's not a lot of code), so this is a suggestion for a `TODO-LLVM`.

Reply:
Yes, ok.  Especially when we get to exceptions.